pkg/operator/ceph/webhook.go

```diff
@@ -30,35 +29,70 @@ import (
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/version"
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
 )
 
 const (
 	admissionControllerAppName       = "rook-ceph-admission-controller"
 	tlsPort                    int32 = 443
+	minK8sVersionWebhook             = "1.16.0"
 )
 
 var (
-	namespace = os.Getenv(k8sutil.PodNamespaceEnvVar)
+	namespace              = os.Getenv(k8sutil.PodNamespaceEnvVar)
+	certManagerWebhookName = "cert-manager-webhook"
 )
 
 func isSecretPresent(ctx context.Context, context *clusterd.Context) (bool, error) {
+	k8sVersion, err := k8sutil.GetK8SVersion(context.Clientset)
+	if err != nil {
+		logger.Errorf("failed to get k8s version. %v", err)
+		return false, nil
+	}
+	if !k8sVersion.AtLeast(version.MustParseSemantic(minK8sVersionWebhook)) {
+		logger.Errorf("minimum versions %s required for webhook to run", err)
+		return false, nil
+	}
+	logger.Infof("Fetching webhook %s to see if cert-manager is installed.", certManagerWebhookName)
+	_, err = context.Clientset.AdmissionregistrationV1().ValidatingWebhookConfigurations().Get(ctx, certManagerWebhookName, metav1.GetOptions{})
+	if err != nil {
+		logger.Info("failed to get cert manager")
+		return false, nil
+	}
+	issuer, err := createIssuer(ctx, context)
+	if err != nil {
+		logger.Errorf("issuer creation failed %v", err)
+		return false, nil
+	}
+	err = createCertificate(ctx, context, issuer)
+	if err != nil {
+		logger.Errorf("certificate creation failed %v", err)
+		return false, nil
+	}
+
 	logger.Infof("looking for admission webhook secret %q", admissionControllerAppName)
 	s, err := context.Clientset.CoreV1().Secrets(namespace).Get(ctx, admissionControllerAppName, metav1.GetOptions{})
```

Comment:
If this secret is not found, then we should skip creating the issuer and cert above, right? So we could call this first, then only create the issuer and cert when it exists.

Reply:
In order to get secret, cert-manager require this issuers and certificate so we have to call this first.